### PR TITLE
Add seed option

### PR DESCRIPTION
**Pull request by SsnL**
Add random seed option to replicate experiment result. For cleanness, this PR sets torch CPU seed, all torch GPU seeds, numpy seed, and python seed to be equal to the same 32-bit int.

Test plan:
Ran CycleGAN multiple times with same seed on SYNTHIA, and observed that results are identical.

### CODE REVIEW

**Comment by hazirbas** [comment]
Did you have any problem regarding the initialization of the weights and biases for normalization? I tried this for pix2pix and setting the seed did not make it reproducable due to using different mean and variance in the ..data.weight.normal_() function:

`def weights_init(m):
    classname = m.__class__.__name__
    if classname.find('Conv') != -1:
        m.weight.data.normal_(0.0, 0.02)
    elif classname.find('BatchNorm2d') != -1:
        m.weight.data.normal_(1.0, 0.02)
        m.bias.data.fill_(0)
`

Moreover, random.seed setting did not set the seed and torchvision.transforms were different at each run.

**Comment by SsnL** [comment]
@hazirbas I tried initialization with same seed, the weights are identical. This is likely from a recent pytorch change that added setting GPU seeds in `manual_seed`. I will change the code so it explicitly set GPU seeds in order to not depend on most recent pytorch.

About `torchvision.transforms`, I realize that they do give different results even if we set same seed. Thanks for helping me finding out this bug. I'll research and see if there are any solutions.

**Comment by SsnL** [comment]
Okay, so the reason that `torchvision.transforms` can give different results is the multithread data loading. There is no guarantee on the order that workers finish. I tried setting seed basing on index, which seemed giving consistent data across training sessions. But it will crop the same patch for each image, which is not ideal. Seed option seem impossible to implement without doing some serious change to the code in either this repo or the pytorch repo. So I will close and abandon.

**Comment by hazirbas** [comment]
@SsnL Hey, thanks a lot for your comments. My solution to transforms was to send Lambda function which uses numpy.random. Then it was consistent for me.

**Comment by deepCMR** [comment]
I get reproducibility in both training and test with adding a fixed seed.  I understand that there are a number of factors during training (weight initialization, etc), but what is random in the test phase?

**Comment by junyanz** [comment]
If you use dropout during test time, you will get randomness. If you use some preprocessing options, which involve random cropping, you might get slightly different results.

**Comment by Elina-Chang** [comment]
> What did you do to address the reproducibility in both training and testing?

**Comment by guyleaf** [comment]
> If you use dropout during test time, you will get randomness. If you use some preprocessing options, which involve random cropping, you might get slightly different results.

Hi @junyanz , I think the latter problem can be solved. The PyTorch official documentation mention that we can add random seed in `work_init_fn` function to get reproducibility, [link](https://pytorch.org/docs/stable/notes/randomness.html#dataloader).
